// mongo-seeding/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const extend = require("extend");
const database_1 = require("./database");
/**
 * Stores default configuration for database seeding.
 */
exports.defaultSeederConfig = {
    database: database_1.defaultDatabaseConfigObject,
    databaseReconnectTimeout: 10000,
    dropDatabase: false,
    dropCollections: false,
};
/**
 * Merges configuration for database seeding.
 *
 * @param partial Partial config object. If not specified, returns a default config object.
 * @param previous Previous config object. If not specified, uses a default config object as a base.
 */
exports.mergeSeederConfig = (partial, previous) => {
    const source = previous ? previous : exports.defaultSeederConfig;
    if (!partial) {
        return source;
    }
    const config = {};
    return extend(true, config, source, partial);
};
/**
 * Stores default collection reading configuration values.
 */
exports.defaultCollectionReadingOptions = {
    extensions: ['json', 'js'],
    transformers: [],
};
/**
 * Merges configuration for collection reading.
 *
 * @param partial Partial config object. If not specified, returns a default config object.
 * @param previous Previous config object. If not specified, uses a default config object as a base.
 */
exports.mergeCollectionReadingOptions = (partial, previous) => {
    const source = previous ? previous : exports.defaultCollectionReadingOptions;
    if (!partial) {
        return source;
    }
    const config = {};
    return extend(true, config, source, partial);
};
